File: src/main/java/cspfj/generator/constraint/DiffGeGenerator.java
package cspfj.generator.constraint;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;

import cspfj.constraint.Constraint;
import cspfj.constraint.semantic.Gt;
import cspfj.constraint.semantic.ReifiedConstraint;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.FunctionalConstraint;
import cspom.constraint.GeneralConstraint;

public final class DiffGeGenerator extends AbstractGenerator {

    public DiffGeGenerator(final Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {

        final Constraint generated;
        if (constraint instanceof GeneralConstraint) {
            generated = generateGeneral(constraint);
        } else if (constraint instanceof FunctionalConstraint) {
            generated = generateReified((FunctionalConstraint) constraint);
        } else {
            throw new IllegalArgumentException(constraint + " not supported");
        }

        if (generated == null) {
            return false;
        }

        addConstraint(generated);

        return true;
    }

    private Constraint generateGeneral(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        final Variable v0;
        final Variable v1;
        final Variable bound;
        Preconditions.checkArgument(
                "diffGe".equals(constraint.getDescription()),
                "Cannot handle %s", constraint);

        v0 = getSolverVariable(constraint.getVariable(0));
        v1 = getSolverVariable(constraint.getVariable(1));
        bound = getSolverVariable(constraint.getVariable(2));

        if (Iterables.any(ImmutableList.of(v0, v1, bound), NULL_DOMAIN)
                || bound.getDomainSize() != 1) {
            return null;
        }

        return new Gt(v0, -bound.getValue(0), v1, false);
    }

    private Constraint generateReified(final FunctionalConstraint constraint)
            throws FailedGenerationException {
        final Variable v0;
        final Variable v1;
        final Variable bound;
        Preconditions.checkArgument(
                "diffGe".equals(constraint.getDescription()),
                "Cannot handle %s", constraint);

        v0 = getSolverVariable(constraint.getVariable(1));
        v1 = getSolverVariable(constraint.getVariable(2));
        bound = getSolverVariable(constraint.getVariable(3));

        if (Iterables.any(ImmutableList.of(v0, v1, bound), NULL_DOMAIN)
                || bound.getDomainSize() != 1) {
            return null;
        }

        final Variable result = getSolverVariable(constraint
                .getResultVariable());
        booleanDomain(result);
        return new ReifiedConstraint(result, new Gt(v0, -bound.getValue(0), v1,
                false), new Gt(v1, bound.getValue(0), v0, true));
    }

}

File: src/main/scala/cspfj/generator/constraint/DiffGeGenerator.scala
package cspfj.generator.constraint;

import cspfj.constraint.semantic.{ ReifiedConstraint, Gt }
import cspfj.{ Variable, Problem }
import cspom.constraint.{ GeneralConstraint, FunctionalConstraint, CSPOMConstraint }

final class DiffGeGenerator(problem: Problem) extends AbstractGenerator(problem) {

  override def generateGeneral(constraint: GeneralConstraint) = {
    require("diffGe" == constraint.description, "Cannot handle" + constraint)
    val Seq(v0, v1, bound) = constraint.scope map cspom2cspfj

    if (bound.dom.size != 1 || Seq(v0, v1, bound).exists(_.dom == null)) {
      false
    } else {
      addConstraint(new Gt(v0, -bound.dom.firstValue, v1, false))
      true
    }
  }

  override def generateFunctional(constraint: FunctionalConstraint) = {
    require("diffGe" == constraint.description, "Cannot handle" + constraint)
    val Seq(result, v0, v1, bound) = constraint.scope map cspom2cspfj

    if (bound.dom.size != 1 || Seq(v0, v1, bound).exists(_.dom == null)) {
      false
    } else {
      AbstractGenerator.booleanDomain(result)
      addConstraint(new ReifiedConstraint(
        result,
        new Gt(v0, -bound.dom.firstValue, v1, false),
        new Gt(v1, bound.dom.firstValue, v0, true)))
      true
    }

  }

}
